### PR TITLE
Add backup_plan and backup_plan_info modules

### PR DESCRIPTION
**Pull request by krisek**
##### SUMMARY
These three modules add capability to manage all configurations related to backups on AWS.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
backup_vault
backup_plan
backup_selection

##### ADDITIONAL INFORMATION

We've been using these three modules for 6 months now in our AWS estate, I thought it might be useful for others.

I know that this is not the repo for public contribution, but as there is already work being done on this area here, I think it still make sense to have a look on the code, I tried to be as comprehensive as possible.

No worries if it doesn't make into the main branch, feel free to cherry pick parts of it.

### CODE REVIEW

**Review by tremble — CHANGES_REQUESTED** [review]
Thanks for this.  I've taken a quick first look, and most of the things that are standing out to me are minor rather than something significant.

You'll see there are multiple layers of linter and tests complaining about things.

The "Sanity" tests 
```
ERROR: Found 13 validate-modules issue(s) which need to be resolved:
ERROR: plugins/modules/backup_plan.py:0:0: import-error: Exception attempting to import module for argument_spec introspection, 'No module named 'ansible.module_utils.resource_tags''
ERROR: plugins/modules/backup_plan.py:0:0: parameter-list-no-elements: DOCUMENTATION.options.advanced_backup_settings: Argument defines type as list but elements is not defined for dictionary value @ data['options']['advanced_backup_settings']. Got {'description': ['Specifies a list of BackupOptions for each resource type. These settings are only available for Windows Volume Shadow Copy Service (VSS) backup jobs.'], 'required': False, 'type': 'list'}
ERROR: plugins/modules/backup_plan.py:0:0: parameter-list-no-elements: DOCUMENTATION.options.rules: Argument defines type as list but elements is not defined for dictionary value @ data['options']['rules']. Got {'description': ['An array of BackupRule objects, each of which specifies a scheduled task that is used to back up a selection of resources.'], 'required': False, 'type': 'list'}
ERROR: plugins/modules/backup_plan.py:5:0: import-before-documentation: Import found before documentation variables. All imports must appear below DOCUMENTATION/EXAMPLES/RETURN.
ERROR: plugins/modules/backup_selection.py:0:0: import-error: Exception attempting to import module for argument_spec introspection, 'No module named 'ansible.module_utils.resource_tags''
ERROR: plugins/modules/backup_selection.py:0:0: parameter-list-no-elements: DOCUMENTATION.options.list_of_tags: Argument defines type as list but elements is not defined for dictionary value @ data['options']['list_of_tags']. Got {'description': ['A list of conditions that you define to assign resources to your backup plans using tags. Condition operators are case sensitive.'], 'required': False, 'type': 'list'}
ERROR: plugins/modules/backup_selection.py:0:0: parameter-list-no-elements: DOCUMENTATION.options.not_resources: Argument defines type as list but elements is not defined for dictionary value @ data['options']['not_resources']. Got {'description': ['A list of Amazon Resource Names (ARNs) to exclude from a backup plan. The maximum number of ARNs is 500 without wildcards, or 30 ARNs with wildcards. If you need to exclude many resources from a backup plan, consider a different resource selection strategy, such as assigning only one or a few resource types or refining your resource selection using tags.'], 'required': False, 'type': 'list'}
ERROR: plugins/modules/backup_selection.py:0:0: parameter-list-no-elements: DOCUMENTATION.options.resources: Argument defines type as list but elements is not defined for dictionary value @ data['options']['resources']. Got {'description': ['A list of Amazon Resource Names (ARNs) to assign to a backup plan. The maximum number of ARNs is 500 without wildcards, or 30 ARNs with wildcards. If you need to assign many resources to a backup plan, consider a different resource selection strategy, such as assigning all resources of a resource type or refining your resource selection using tags.'], 'required': False, 'type': 'list'}
ERROR: plugins/modules/backup_selection.py:0:0: return-syntax-error: RETURN.backup_selection.contains.creation_date.sample: not a valid value for dictionary value @ data['backup_selection']['contains']['creation_date']['sample']. Got datetime.datetime(2023, 1, 24, 10, 8, 3, 193000, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600)))
ERROR: plugins/modules/backup_selection.py:5:0: import-before-documentation: Import found before documentation variables. All imports must appear below DOCUMENTATION/EXAMPLES/RETURN.
ERROR: plugins/modules/backup_vault.py:0:0: import-error: Exception attempting to import module for argument_spec introspection, 'No module named 'ansible.module_utils.resource_tags''
ERROR: plugins/modules/backup_vault.py:0:0: return-syntax-error: RETURN.backup_vault.contains.tags.contains: required key not provided @ data['backup_vault']['contains']['tags']['contains']. Got None
ERROR: plugins/modules/backup_vault.py:6:0: import-before-documentation: Import found before documentation variables. All imports must appear below DOCUMENTATION/EXAMPLES/RETURN.
```

The 'black' linter:
https://github.com/ansible-collections/amazon.aws/actions/runs/4541478098/jobs/8003940319?pr=1446

I've also picked out a few other things.

One minor additional thing I'd usually tweak:  It's much easier to follow the code if you break it out into functions: one big "main" with lots of nested statements can be kinda hard to follow.

If you're willing to make these changes that'd be great.  You can find some more documentation about writing tests in our docs https://github.com/ansible-collections/amazon.aws/blob/main/docs/docsite/rst/dev_guidelines.rst#integration-tests-for-aws-modules

File: plugins/modules/backup_plan.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
+#!/usr/bin/python
+# Copyright: Ansible Project
```
[the remark this answers is not in the document]
Comment (krisek) [reply plugins/modules/backup_plan.py]: fixed in ca3c32ac

File: plugins/modules/backup_plan.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+from __future__ import absolute_import, division, print_function
```
[the remark this answers is not in the document]
Comment (krisek) [reply plugins/modules/backup_plan.py]: fixed in ca3c32ac

File: plugins/modules/backup_plan.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+from ansible_collections.amazon.aws.plugins.module_utils.ec2 import boto3_tag_list_to_ansible_dict
+from ansible_collections.amazon.aws.plugins.module_utils.ec2 import ansible_dict_to_boto3_filter_list
+from ansible_collections.amazon.aws.plugins.module_utils.ec2 import AWSRetry
+from ansible_collections.amazon.aws.plugins.module_utils.core import is_boto3_error_code
+from ansible_collections.amazon.aws.plugins.module_utils.core import AnsibleAWSModule
+__metaclass__ = type
+DOCUMENTATION = '''
+module: backup_plan
+short_description: create, delete backup plans
+version_added: 1.0.0
```
[the remark this answers is not in the document]
Comment (krisek) [reply plugins/modules/backup_plan.py]: fixed in ca3c32ac

File: plugins/modules/backup_plan.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+from __future__ import absolute_import, division, print_function
+import json
+from ansible.module_utils.resource_tags import ensure_tags
+from ansible_collections.amazon.aws.plugins.module_utils.ec2 import camel_dict_to_snake_dict
+from ansible_collections.amazon.aws.plugins.module_utils.ec2 import boto3_tag_list_to_ansible_dict
+from ansible_collections.amazon.aws.plugins.module_utils.ec2 import ansible_dict_to_boto3_filter_list
+from ansible_collections.amazon.aws.plugins.module_utils.ec2 import AWSRetry
+from ansible_collections.amazon.aws.plugins.module_utils.core import is_boto3_error_code
+from ansible_collections.amazon.aws.plugins.module_utils.core import AnsibleAWSModule
+__metaclass__ = type
```
[the remark this answers is not in the document]
Comment (krisek) [reply plugins/modules/backup_plan.py]: fixed in ca3c32ac

File: plugins/modules/backup_plan.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+DOCUMENTATION = '''
+module: backup_plan
+short_description: create, delete backup plans
```
[the remark this answers is not in the document]
Comment (krisek) [reply plugins/modules/backup_plan.py]: fixed in ca3c32ac

File: plugins/modules/backup_plan.py [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+    choices: ['present', 'absent']
+    type: str
+notes: []
+author:
+  - Kristof Imre Szabo (@krisek)
+extends_documentation_fragment:
+  - amazon.aws.aws
+  - amazon.aws.ec2
+  - amazon.aws.boto3
+  - amazon.aws.tags
```
[the remark this answers is not in the document]
Comment (krisek) [reply plugins/modules/backup_plan.py]: fixed in ca3c32ac

File: plugins/modules/backup_plan.py [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+import json
+from ansible.module_utils.resource_tags import ensure_tags
+from ansible_collections.amazon.aws.plugins.module_utils.ec2 import camel_dict_to_snake_dict
+from ansible_collections.amazon.aws.plugins.module_utils.ec2 import boto3_tag_list_to_ansible_dict
+from ansible_collections.amazon.aws.plugins.module_utils.ec2 import ansible_dict_to_boto3_filter_list
+from ansible_collections.amazon.aws.plugins.module_utils.ec2 import AWSRetry
+from ansible_collections.amazon.aws.plugins.module_utils.core import is_boto3_error_code
+from ansible_collections.amazon.aws.plugins.module_utils.core import AnsibleAWSModule
+__metaclass__ = type
+DOCUMENTATION = '''
```
[the remark this answers is not in the document]
Comment (krisek) [reply plugins/modules/backup_plan.py]: fixed in ca3c32ac

File: plugins/modules/backup_plan.py [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+
+
+def main():
+    argument_spec = dict(
+        backup_plan_name=dict(type='str', required=True),
+        rules=dict(type='list', required=False),
+        advanced_backup_settings=dict(type='list', required=False),
+        tags=dict(required=False, type='dict', aliases=['resource_tags']),
+        purge_tags=dict(default=True, type='bool'),
+        state=dict(default='present', choices=['present', 'absent'])
```
[the remark this answers is not in the document]
Comment (krisek) [reply plugins/modules/backup_plan.py]: fixed in ca3c32ac

File: plugins/modules/backup_plan.py [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+    elif state == 'absent':
+
+        if exist:
+            try:
+                response_delete = client.delete_backup_plan(
+                    aws_retry=True, BackupPlanId=response.get('BackupPlanId'))
+                if (response_delete['ResponseMetadata']['HTTPStatusCode'] == 200):
+                    changed = True
+            except Exception as e:
+                module.exit_json(changed=changed, failed=True)
```
[the remark this answers is not in the document]
Comment (krisek) [reply plugins/modules/backup_plan.py]: for plan and vault fixed in ca3c32ac -- selection is to be done

**Comment by krisek** [comment]
Thanks for reviewing it so quickly, I fixed most of the things. Breaking to functions will need to take a deeper breath, maybe it would make sense to do after test cases are there, so that we can see eventual regressions immediately. 

I still see a conflict though with #1436 

How should we approach this problem?

**Comment by tremble** [comment]
> Thanks for reviewing it so quickly, I fixed most of the things. Breaking to functions will need to take a deeper breath, maybe it would make sense to do after test cases are there, so that we can see eventual regressions immediately.

That's fine.  You've seen the other PRs, does the way we write our integration tests make sense to you?  (they're essentially just roles in tests/integration/targets)
 
> I still see a conflict though with #1436
> 
> How should we approach this problem?

I'd probably drop the backup_vault module from this PR, since @GomathiselviS is paid to work on this she's probably likely to be ready with her PR before we could get these merged.

File: plugins/modules/backup_plan.py [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+  ansible.builtin.debug:
+    var: elastic_plan
+"""
+RETURN = """
+backup_plan_name:
+  description: backup plan name
+  returned: always
+  type: str
+  sample: elastic
+backup_plan_id:
```
Comment (GomathiselviS) [inline plugins/modules/backup_plan.py:79]: I donot see backup_plan_name and backup_plan_id as part of the response structure in the [documentation](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/backup/client/create_backup_plan.html). Are you adding these keys to the return dict?

File: plugins/modules/backup_plan.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (krisek) [reply plugins/modules/backup_plan.py]: fixed with 5dc8e2d5

File: plugins/modules/backup_plan.py (hunk 10 above)
Comment (krisek) [reply plugins/modules/backup_plan.py:79]: many thanks, very good catch, it really returned a `null` when creating a plan, fixed with 99b2dabc

**Comment by krisek** [comment]
> Are you planning to add backup_plan_info and backup_selection_info modules? If not I can start with those modules.

Thanks a lot, I started to work on backup_plan_info based on yours... got stuck a bit with tag retrieval. But it's progressing.

**Comment by gravesm** [comment]
@krisek Thanks for all your work on these modules. If you have work that hasn't been pushed yet, feel free to add it. We'll likely try to pick this up and continue on what you've already done, as we'd like to get these into the 6.0.0 release coming up in a couple weeks.

**Comment by krisek** [comment]
there is one info module missing and tests cases of course, I'll take care about the info module in the next two days

**Comment by alinabuzachis** [comment]
> there is one info module missing and tests cases of course, I'll take care about the info module in the next two days

@krisek Thank your for this contribution. If you need any help, just let us know here or on IRC Libera.Chat #ansible-aws.

**Comment by hakbailey** [comment]
@krisek I just joined the team at Ansible and can start work on the tests for these modules. Let me know ASAP if you've already made progress on those though as I don't want to duplicate your work.

**Comment by krisek** [comment]
just starting to create the missing info module, sorry this week was more turbulent than I expected

**Comment by alinabuzachis** [comment]
I have split this PR in two because backup_selection modules still require some re-work. The PR containing only backup_plan* modules is this one https://github.com/ansible-collections/amazon.aws/pull/1523

**Comment by tremble** [comment]
recheck

File: plugins/modules/backup_plan_info.py [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+  - Kristof Imre Szabo (@krisek)
+  - Alina Buzachis (@alinabuzachis)
+options:
+  backup_plan_names:
+    type: list
+    elements: str
+    default: []
+    description:
+      - Specifies a list of plan names.
+      - If an empty list is specified, information for the backup plans in the current region is returned.
```
Comment (alinabuzachis) [inline plugins/modules/backup_plan_info.py:26]: I guess we should add note like "Listing all the AWS backup plans is not supported at the moment. When I(backup_plan_names) is not set, an empty list is returned." We can then add this functionality next.

Comment (alinabuzachis) [reply plugins/modules/backup_plan_info.py:26]: @tremble What do you suggest?

Comment (tremble) [reply plugins/modules/backup_plan_info.py:26]: Keep it simple.  Make it required for now, we can make it optional later as a "new feature" (non-breaking IMO).

Comment (alinabuzachis) [reply plugins/modules/backup_plan_info.py:26]: Right, no breaking changes! Makes sense.

Comment (alinabuzachis) [reply plugins/modules/backup_plan_info.py:26]: My bad! I guess that functionality works https://github.com/ansible-collections/amazon.aws/pull/1446/files#diff-d2301559aaf1b18099212d421d58fc78b99063e3f730e0c23c1cc79cff553402R114-R121 (missed this part), @hakbailey can you please confirm?

File: changelogs/fragments/1446-backup_plan-add-new-module.yml [hunk 12]
```diff
@@ -0,0 +1,3 @@
+trivial:
```
Comment (GomathiselviS) [inline changelogs/fragments/1446-backup_plan-add-new-module.yml:1]: changelog is not necessary when we add new modules. It gets added to the release summary during the release.

Comment (tremble) [reply changelogs/fragments/1446-backup_plan-add-new-module.yml:1]: The tests have been inconsistent in handling this :)  trivial can stay, just for the sake of not tripping over a test

Comment (hakbailey) [reply changelogs/fragments/1446-backup_plan-add-new-module.yml:1]: Yeah sanity tests failed without it

**Review by alinabuzachis — APPROVED** [review]

File: changelogs/fragments/1446-backup_plan-add-new-module.yml (hunk 12 above)
Comment (hakbailey) [reply changelogs/fragments/1446-backup_plan-add-new-module.yml:1]: Or um, maybe not sanity tests....I'm still learning what all the various checks are. But _something_ was failing :)

**Review by GomathiselviS — APPROVED** [review]

**Comment by alinabuzachis** [comment]
regate
